library/core/src/any.rs

```diff
@@ -949,6 +1011,93 @@ impl<'a> Demand<'a> {
         }
         self
     }
+
+    /// Check if the `Demand` would be satisfied if provided with a
+    /// value of the specified type. If the type does not match or has
+    /// already been provided, returns false.
+    ///
+    /// # Examples
+    ///
+    /// Check if an `u8` still needs to be provided and then provides
+    /// it.
+    ///
+    /// ```rust
+    /// #![feature(provide_any)]
+    ///
+    /// use std::any::{Provider, Demand};
+    ///
+    /// struct SomeConcreteType(Option<u8>);
+    ///
+    /// impl Provider for SomeConcreteType {
+    ///     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    ///         if let Some(v) = self.0 {
+    ///             demand.provide_value::<u8>(v);
+    ///         }
+    ///
+    ///         assert!(
+    ///             !demand.would_be_satisfied_by_value_of::<u8>(),
+    ///             "We know an u8 has already been provided",
+    ///         );
```

Comment:
I've placed a complementary pair of asserts inside the `if let`, but that made me wonder if there's a _point_ to the `Option` in this example now... thoughts?

Reply:
I think you're right, we got a little away from the original point of the example, I'm thinking we need to reintroduce some "expensive operation". What if instead of an `Option` we take a closure that returns an option and call the member "extremely_expensive_operation" or something fun along those lines, then we check if the demand would be satisfied by the `T` type of the option before the if-let, then assert that it no longer can be fulfilled after the provide within the if-let.

Reply:
What do you think about 


```rust
struct Parent(Option<u8>);

impl Provider for Parent {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
        if let Some(v) = self.0 {
            demand.provide_value::<u8>(v);
        }
    }
}

struct Child {
    parent: Parent,
    we_expect_to_run_the_expensive_computation: bool,
}

impl Child {
    // Pretend that this takes a lot of resources to evaluate.
    fn an_expensive_computation(&self) -> Option<u8> {
        Some(99)
    }
}

impl Provider for Child {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
        // In general, we don't know if this call will provide
        // an `u8` value or not...
        self.parent.provide(demand);

        // ...so we check to see if the `u8` is needed before
        // we run our expensive computation.
        if demand.would_be_satisfied_by_value_of::<u8>() {
            assert!(self.we_expect_to_run_the_expensive_computation);

            if let Some(v) = self.an_expensive_computation() {
                demand.provide_value::<u8>(v);
            }
        }
    }
}

let parent = Parent(Some(42));
let child = Child { parent, we_expect_to_run_the_expensive_computation: false };
assert_eq!(Some(42), std::any::request_value::<u8>(&child));

let parent = Parent(None);
let child = Child { parent, we_expect_to_run_the_expensive_computation: true };
assert_eq!(Some(99), std::any::request_value::<u8>(&child));
```

Reply:
This looks good, I think you're right on needing to have a chain of provides via the parent child relationship.

My one confusion is around the `we_expect_to_run_the_expensive_computation`. I'm not sure what the point behind this is and what it adds to this example, and I worry that it makes it harder to focus on the relevant parts. Do you think it's important to keep this or do you think it's fine to remove?

Reply:
> what the point behind this is and what it adds to this example

I believe if we removed that assertion, then the implementations of `would_be_satisfied_by_*` could be replaced with a hard-coded `true` and we'd not notice because of the behavior to only keep the first provided value. However, the expensive computation would still be executed, negating the point of the `would_be_satisfied_by_*` methods.

Reply:
I see, I wouldn't have thought to defensively code against such an implementation of `would_be_satisfied_by_*`. I think we can get a similar benefit without having to do the same manual tracking by asserting that `would_be_satisfied_by` is false after the second provide.

Reply:
Good call. Updated and rebased for good measure.